Src/Extern/SUNDIALS/AMReX_SUNMemory.cpp

```diff
@@ -0,0 +1,148 @@
+#include <AMReX.H>
+#include <AMReX_Gpu.H>
+#include <AMReX_SUNMemory.H>
+#if defined(AMREX_USE_HIP)
+#include <sunmemory/sunmemory_hip.h>
+#elif defined(AMREX_USE_CUDA)
+#include <sunmemory/sunmemory_cuda.h>
+#elif defined(AMREX_USE_DPCPP)
+#include <sunmemory/sunmemory_sycl.h>
+#endif
+
+namespace amrex {
+namespace sundials {
+
+namespace {
+  int Alloc(SUNMemoryHelper, SUNMemory* memptr, size_t memsize, SUNMemoryType mem_type)
+  {
+    SUNMemory mem = SUNMemoryNewEmpty();
+
+    mem->ptr = NULL;
+    mem->own = SUNTRUE;
+    mem->type = mem_type;
+
+    if (mem_type == SUNMEMTYPE_HOST) {
+      mem->ptr = The_Cpu_Arena()->alloc(memsize);
+    } else if (mem_type == SUNMEMTYPE_UVM) {
+        if (The_Arena()->isManaged()) {
+          mem->ptr = The_Arena()->alloc(memsize);
+        } else if (The_Managed_Arena()->isManaged()) {
+          mem->ptr = The_Managed_Arena()->alloc(memsize);
+        } else {
+          free(mem);
+          return -1;
+        }
+    } else if (mem_type == SUNMEMTYPE_DEVICE) {
+      mem->ptr = The_Arena()->alloc(memsize);
+    } else if (mem_type == SUNMEMTYPE_PINNED) {
+      mem->ptr = The_Pinned_Arena()->alloc(memsize);
+    } else {
+      free(mem);
+      return -1;
+    }
+
+    *memptr = mem;
+    return 0;
+  }
+
+  int Dealloc(SUNMemoryHelper, SUNMemory mem)
```

Comment:
This function `Dealloc` needs to use the same Arena as `Alloc`.  If they do not match, there will be a runtime error.  Maybe we can put the logic in a function say
```
Arena* getArena (SUNMemoryType type)
{
    if (type == ...) {
        return ....;
    } else if (type == ...) {
    } ....
    else {
        return nullptr;
    } 
}
```

Then `Alloc` can use it and this function will be
```
if (mem == nullptr) return 0;
auto arena = getArena(mem->type);
if (arena) {
    arena->free(mem->ptr);
    free(mem);
    return 0;
else {
    free(mem);
    return -1;
}
}
```

Reply:
I get a runtime error when I use the current code with HIP, because of the Dealloc mismatch with The_Device_Arena and the isManaged(). When I duplicate the logic from Alloc into Dealloc, this runtime error goes away, and the Nyx gpu_feature branch built with GMake works for Cuda, Hip, and Sycl. Having a separate getArena function would be great.

Reply:
@jmsexton03 Can you update the PR with the fixes you have described?

Reply:
@balos1  I submitted a PR to this branch [https://github.com/balos1/amrex/pull/2](url) 
It includes the initial commit I referenced earlier, my unsuccessful attempt to tweak cmake to work for a Nyx superbuild, and the final commit for the getArena implementation. I'll go in and fix the spacing and push again, but I don't know yet if the other CI tests will pass.

Reply:
@jmsexton03 OK, I merged that PR. Just waiting to see if the CI tests pass here.

Reply:
@balos1 Looks like it passed.
I'm testing the Nyx CI with this PR in the Nyx branch `gpu_feature`, which is ready to become Nyx `development` 
https://github.com/AMReX-Astro/Nyx/commit/636371e2df64f13fbe300422e3c0acec9004c143